### DISCUSSION TITLE
Software only works on CPU

### DISCUSSION THREAD

**Discussion by Mousa-saleh** [neuropoly/axondeepseg#560] [Bug reports, suggestions for code refactoring]
Hello all, 

I have a laptop with RTX 3060 GPU.
The software was installed and its working but cannot integrate it with the GPU, here is what I did.

1) installed visual studio 2019.
2)downloaded and installed cuda 10,  and cudnn 7.4 and added it to path.
3)   wrote inside miniconda enviroment (conda install -c anaconda tensorflow-gpu==1.13.1), 
then (conda install -c conda-forge keras==2.2.4) 

Can you tell what is missing here,

Thanks,

**Comment by Mousa-saleh** [comment]
From what I have read, RTX series only works with CUDA11.x. This means I have to use tensorflow 2. 
When did that, GPU was detected but got lots of errors, maybe bacause the software is written on older versions of tensorflow and keras.

Is there a way to circumvent this problem??

Thanks,

**Comment by mathieuboudreau** [comment]
Hi @Mousa-saleh,

Yes you are correct, to work with Cuda 11.x, we would need to upgrade the Tensorflow version from 1.x to 2.x.

Unfortunately, this is a major upgrade in Tensorflow versions, and thus the API changes significantly, so it would require significant changes in our code to make it compatible with Tensorflow 2.x. Simply changing the version in our environment file would cause errors, as you experienced.

However, we are currently in the process of changing how we handle the machine learning backend of AxonDeepSeg to start using IVADOMED, see PR #547. I think (and maybe @mariehbourget can confirm) that you would be able to use Cuda 11.x after this upgrade, which is expected to be done in a month or two.

**Comment by Mousa-saleh** [comment]
OK, l see.

Do you think it can work if l move the software into colab??

Thanks,

**Reply by mathieuboudreau** [reply]
Hi @Mousa-saleh,

We touched briefly on this topic in another thread you had opened, see here: https://github.com/neuropoly/axondeepseg/discussions/526#discussioncomment-875560

I don't know if there's anything limiting the use of ADS on Google Colab, however I don't know anyone who's dedicated sufficient time to try. If you do, please report back on your experience.

Thanks!

Mathieu

**Reply by Mousa-saleh** [reply]
Ok, 
I will give it a shot and see what I get.

Thanks,

**Comment by Stoyan-I-A** [comment]
> From what I have read, RTX series only works with CUDA11.x. This means I have to use tensorflow 2.

Are you sure? I tested the GPU installation with a RTX 2070 a while ago and it was working. 
EDIT: I didn't train a model however, I just segmented an image using the default model. You might also be right and the tensorflow version we're using isn't working on the RTX 3000 series.

> wrote inside miniconda enviroment (conda install -c anaconda tensorflow-gpu==1.13.1),

Did you uninstall the regular tensorflow beforehand? If both are present, the regular tensorflow will be used instead of tensorflow-gpu.

We have some instructions on our documentation page about the [GPU compatible installation](https://axondeepseg.readthedocs.io/en/latest/documentation.html#gpu-compatible-installation).

**Comment by Mousa-saleh** [comment]
Can you tell me what Cuda version you installed??

**Reply by Stoyan-I-A** [reply]
Using this method, I'm not sure how it works but by using our installation procedure and installing AxonDeepSeg in a Conda environment, Conda seems to install CUDA 10.0 in the environment as well as cuDNN 7.4.

That being said, I do also have CUDA 10.1 installed on my machine but this one is not compatible with tensorflow-gpu 1.13.1.

Also, I don't know if you saw the edit I made but it is possible that you are right and the tensorflow version we're using isn't working on the RTX 3000 series, I just don't have a machine with that kind of GPU to test it.

**Comment by Mousa-saleh** [comment]
Is it safe for the laptop to run the software on the CPU, also, what is the approximate time for the new version of ads to be available??

Thanks,

**Reply by Stoyan-I-A** [reply]
> Is it safe for the laptop to run the software on the CPU

That is a good question. The main concern would be overheating during the training. People on the internet have been training DL models (of all kinds, not just AxonDeepSeg) on laptops and reported high temperatures, which is expected since laptops have limited space for heat sinks. 

That being said, that doesn't mean that the temperatures are high enough to cause damage. If I had to answer with a yes or a no, I would say yes, it should be safe. 

Note that the CPU would be loaded for a long period of time even if you run a training using the GPU.

As for your second question, I will let the rest of the team answer it.

**Comment by Mousa-saleh** [comment]
OK,

Thank you very much.

**Comment by mariehbourget** [comment]
Hi @Mousa-saleh,

The training of models in the new version of ADS will take place in `ivadomed` (https://ivadomed.org/). Then, the new trained model will be applied for segmentation and morphometrics calculation in ADS.

Training on microscopy images in `ivadomed`, as well as compatibility with Cuda 11.x, are being implemented at the moment and should be ready in a month or two.

Best,
Marie-Hélène

**Comment by Mousa-saleh** [comment]
Great, thanks for the update.
